File: agentflow/server/models/graph.py
from pydantic import BaseModel, Field
from typing import Optional


class GraphNodeData(BaseModel):
    label: str
    type: str  # input | orchestrator | agent | checkpoint | output
    status: str = "pending"
    agent_id: Optional[str] = None
    agent_color: Optional[str] = None
    agent_icon: Optional[str] = None
    result: Optional[str] = None
    duration: Optional[int] = None
    input_modality: Optional[str] = None
    timestamp: Optional[str] = None


class GraphNode(BaseModel):
    id: str
    type: str
    data: GraphNodeData
    position: dict = Field(default_factory=lambda: {"x": 0, "y": 0})


class GraphEdgeData(BaseModel):
    status: str = "pending"
    data_preview: Optional[str] = None


class GraphEdge(BaseModel):
    id: str
    source: str
    target: str
    data: GraphEdgeData = Field(default_factory=GraphEdgeData)


class ExecutionGraphState(BaseModel):
    task_id: str = ""
    nodes: list[dict] = Field(default_factory=list)
    edges: list[dict] = Field(default_factory=list)
    current_node_id: Optional[str] = None
    status: str = "planning"

File: agentflow/client/src/types/graph.ts
import { type Node, type Edge } from '@xyflow/react';

export type GraphNodeType = 'input' | 'orchestrator' | 'agent' | 'checkpoint' | 'output';

export type NodeStatus = 'pending' | 'running' | 'completed' | 'failed' | 'awaiting_approval' | 'approved' | 'skipped';

export type EdgeStatus = 'pending' | 'active' | 'completed' | 'failed';

export interface GraphNodeData extends Record<string, unknown> {
  label: string;
  type: GraphNodeType;
  status: NodeStatus;
  agentId?: string;
  agentColor?: string;
  agentIcon?: string;
  result?: string;
  duration?: number;
  inputModality?: 'text' | 'voice' | 'image';
  timestamp?: string;
}

export interface GraphEdgeData extends Record<string, unknown> {
  status: EdgeStatus;
  dataPreview?: string;
}

export type FlowNode = Node<GraphNodeData>;
export type FlowEdge = Edge<GraphEdgeData>;

export interface ExecutionGraphState {
  taskId: string;
  nodes: FlowNode[];
  edges: FlowEdge[];
  currentNodeId?: string;
  status: 'planning' | 'awaiting_approval' | 'executing' | 'completed' | 'failed';
}

export type ExecutionSSEEvent =
  | { type: 'graph_init'; graph: ExecutionGraphState }
  | { type: 'node_status'; nodeId: string; status: NodeStatus; result?: string; duration?: number }
  | { type: 'edge_status'; edgeId: string; status: EdgeStatus; dataPreview?: string }
  | { type: 'checkpoint_reached'; nodeId: string; stepId: string }
  | { type: 'execution_complete'; graph: ExecutionGraphState; summary?: string }
  | { type: 'execution_failed'; nodeId: string; error: string };
